pkg/tiles/tile.go

```diff
@@ -0,0 +1,90 @@
+package tiles
+
+import (
+	buildings "github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/tiles/buildings"
+	connectionMod "github.com/YetAnotherSpieskowcy/Carcassonne-Engine/pkg/tiles/connection"
+)
+
+/*
+Immutable object
+*/
+type Tile struct {
+	Features  []Feature
+	HasShield bool
+	Building  buildings.Bulding
+}
+
+func (tile *Tile) Cities() [][]connectionMod.Side {
+	var cities [][]connectionMod.Side
+	for _, feature := range tile.Features {
+		if feature.FeatureType == CITY {
+			cities = append(cities, feature.Connections)
+		}
+	}
+	return cities
+}
+
+func (tile *Tile) CitiesAppendConnection(connections []connectionMod.Side) {
+	tile.Features = append(tile.Features, Feature{
+		FeatureType: CITY,
+		Connections: connections,
+	})
+
+}
+
+func (tile *Tile) Roads() [][]connectionMod.Side {
+	var roads [][]connectionMod.Side
+	for _, feature := range tile.Features {
+		if feature.FeatureType == ROAD {
+			roads = append(roads, feature.Connections)
+		}
+	}
+	return roads
+}
+
+func (tile *Tile) RoadsAppendConnection(connections []connectionMod.Side) {
+	tile.Features = append(tile.Features, Feature{
+		FeatureType: ROAD,
+		Connections: connections,
+	})
+}
+
+func (tile *Tile) Fields() [][]connectionMod.Side {
+	var fields [][]connectionMod.Side
+	for _, feature := range tile.Features {
+		if feature.FeatureType == FIELD {
+			fields = append(fields, feature.Connections)
+		}
+	}
+	return fields
+}
+
+func (tile *Tile) FieldsAppendConnection(connections []connectionMod.Side) {
+	tile.Features = append(tile.Features, Feature{
+		FeatureType: FIELD,
+		Connections: connections,
+	})
+}
+
+func (tile Tile) Rotate(rotations uint) Tile {
+	var t Tile
+	// rotate cities
+	for _, cityConnection := range tile.Cities() {
+		t.CitiesAppendConnection(connectionMod.RotateSideArray(cityConnection, rotations))
+	}
+
+	// rotate roads
+	for _, road := range tile.Roads() {
+		t.RoadsAppendConnection(connectionMod.RotateSideArray(road, rotations))
+	}
+
+	// rotate fields
+	for _, field := range tile.Fields() {
+		t.FieldsAppendConnection(connectionMod.RotateSideArray(field, rotations))
+	}
+
+	// copy parameters
+	t.HasShield = tile.HasShield
+	t.Building = tile.Building
+	return t
```

Comment:
In this case we need to return tile, since otherwise this function becomes a maintenance burden because modifications to Tile structs needs to be reflected in this function. Currently adding new field to the Tile struct will cause resource leak when calling this function.

Reply:
what do you want me to do?

Reply:
> In this case we need to return tile

Reply:
so return the same tile, which was passed as the argument??????

Reply:
Return exact copy of the input tile while only changing tile's connections.
So instead of manually setting every field rely on build in golang value receiver feature.
From https://go.dev/tour/methods/4:
> With a value receiver, the Scale method operates on a copy of the original Vertex value.

Reply:
won't it modyfy the original tile?

Reply:
It won't since tile is not a pointer receiver but a value receiver. I would highly encourage you to take a look at the link i posted in this conversation.

Reply:
check now

Reply:
Looks ok